experimental/trusted_shuffler/trusted_shuffler/src/lib.rs

```diff
@@ -16,4 +16,140 @@
 
 //! Trusted Shuffler implementation.
 
-// TODO(#2614): Implement Trusted Shuffler logic.
+#[cfg(test)]
+mod tests;
+
+use anyhow::{anyhow, Context};
+use async_trait::async_trait;
+use futures::future::join_all;
+use std::{
+    marker::{Send, Sync},
+    mem::replace,
+    ops::DerefMut,
+    sync::{Arc, Mutex},
+};
+use tokio::sync::oneshot;
+
+type Request = Vec<u8>;
+type Response = Vec<u8>;
+
+struct Message {
+    // Determines the original order in which messages arrived.
+    number: usize,
```

Comment:
how about `arrived_as` or `arrival_position`? Maybe we could document what we use this for? Also, is this the order of when the `data` for this message arrived?